extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java

```diff
@@ -146,19 +147,31 @@ protected SegmentReference mapSegment(final Segment segment)
 
   private boolean initializeSegmentMapFn(final IntSet readableInputs)
   {
+    final AtomicLong cpuAccumulator = new AtomicLong();
     if (segmentMapFn != null) {
       return true;
     } else if (broadcastJoinHelper == null) {
       segmentMapFn = Function.identity();
       return true;
     } else {
-      final boolean retVal = broadcastJoinHelper.buildBroadcastTablesIncrementally(readableInputs);
-
-      if (retVal) {
-        segmentMapFn = broadcastJoinHelper.makeSegmentMapFn(query);
+      if (query.getDataSource() instanceof InputNumberDataSource) {
+        final boolean retVal = broadcastJoinHelper.buildBroadcastTablesIncrementally(readableInputs);
+        if (retVal) {
+          InputNumberDataSource inputNumberDataSource = (InputNumberDataSource) query.getDataSource();
+          // The InputNumberDataSource requires a BroadcastJoinHelper to be able to create its
+          // segment map function.  It would be a lot better if the InputNumberDataSource actually
+          // had a way to get that injected into it on its own, but the relationship between these objects
+          // was figured out during a refactor and using a setter here seemed like the least-bad way to
+          // make progress on the refactor without breaking functionality.  Hopefully, some future
+          // developer will move this away from a setter.
+          inputNumberDataSource.setBroadcastJoinHelper(broadcastJoinHelper);
+          segmentMapFn = inputNumberDataSource.createSegmentMapFunction(query, cpuAccumulator);
+        }
+        return retVal;
+      } else {
+        segmentMapFn = Function.identity();
```

Comment:
To continue with Abhishek's point if the parent datasource is a `JoinDataSource` with child data sources being `InputNumberDataSource`, this would basically set the map function to an identity function which is not right. One way is to first construct the data source using the broadcastJoinHelper and call createSegmentMapFunction of that. I'll discuss this with @imply-cheddar today. I believe the else part can be simplified as
```
else {
      final boolean retVal = broadcastJoinHelper.buildBroadcastTablesIncrementally(readableInputs);
      if (retVal) {
        final DataSource dataSourceWithInlinedChannelData = broadcastJoinHelper.inlineChannelData(query.getDataSource());
        segmentMapFn = dataSourceWithInlinedChannelData.createSegmentMapFunction(query, cpuAccumulator);
      }
      else {
        segmentMapFn = Function.identity();
      }
      return true;
```

Reply:
Is there a test for that case?

Reply:
Yes there is a `MSQSelectTest.joinTest`

Reply:
If it works before the code change, then why do we need the code change?

Reply:
What I'm trying to say is, if there's not a test that can exercise the corner that is being mentioned, then the corner doesn't exist.  If the corner does exist, there should be a test that exercises it which is broken.  So, we are in 1 of 3 states:

1) There should be a test that's breaking, but there isn't yet.  In that case, let's create the test and then talk about solutions.
2) There is a test that exercises the corner, but it's passing.  In this case, the corner doesn't actually exist and we do not need further code changes.
3) There is no test that generates the highlighted state because something upstream from this code is actually preventing that state from ever existing.  So, it is impossible to actually get to the bad state.  In this case, the corner doesn't actually exist, no code changes required.

Reply:
This follows case 1. The new test case has been added